lang/test/org/partiql/lang/eval/EvaluatingCompilerExecTests.kt

```diff
@@ -0,0 +1,302 @@
+    }
+}
+
+/**
+ * Simple stored procedure that takes one integer argument and outputs that argument back.
+ */
+class OneArgProcedure(valueFactory: ExprValueFactory): NullPropagatingProcedure("one_arg_procedure", 1, valueFactory) {
+    override fun eval(env: Environment, args: List<ExprValue>): ExprValue {
+        val arg = args.first()
+        if (arg.type != ExprValueType.INT) {
+            val errorContext = createWrongSProcErrorContext(arg, "INT", name)
+            throw EvaluationException("invalid first argument",
+                ErrorCode.EVALUATOR_INCORRECT_TYPE_OF_ARGUMENTS_TO_PROCEDURE_CALL,
+                errorContext,
+                internal = false)
+        }
+        return arg
+    }
+}
+
+/**
+ * Simple stored procedure that takes two integer arguments and outputs the args as a string separated by
+ * a space.
+ */
+private class TwoArgProcedure(valueFactory: ExprValueFactory): NullPropagatingProcedure("two_arg_procedure", 2, valueFactory) {
+    override fun eval(env: Environment, args: List<ExprValue>): ExprValue {
+        val arg1 = args.first()
+        if (arg1.type != ExprValueType.INT) {
+            val errorContext = createWrongSProcErrorContext(arg1, "INT", name)
+            throw EvaluationException("invalid first argument",
+                ErrorCode.EVALUATOR_INCORRECT_TYPE_OF_ARGUMENTS_TO_PROCEDURE_CALL,
+                errorContext,
+                internal = false)
+        }
+
+        val arg2 = args[1]
+        if (arg2.type != ExprValueType.INT) {
+            val errorContext = createWrongSProcErrorContext(arg2, "INT", name)
+            throw EvaluationException("invalid second argument",
+                ErrorCode.EVALUATOR_INCORRECT_TYPE_OF_ARGUMENTS_TO_PROCEDURE_CALL,
+                errorContext,
+                internal = false)
+        }
+        return valueFactory.newString("$arg1 $arg2")
+    }
+}
+
+/**
+ * Simple stored procedure that takes one string argument and checks if the binding (case-insensitive) is in the
+ * current environment. If so, returns the value associated with that binding. Otherwise, returns missing.
+ */
+private class OutputBindingProcedure(valueFactory: ExprValueFactory): NullPropagatingProcedure("output_binding", 1, valueFactory) {
+    override fun eval(env: Environment, args: List<ExprValue>): ExprValue {
+        val arg = args.first()
+        if (arg.type != ExprValueType.STRING) {
+            val errorContext = createWrongSProcErrorContext(arg, "STRING", name)
+            throw EvaluationException("invalid first argument",
+                ErrorCode.EVALUATOR_INCORRECT_TYPE_OF_ARGUMENTS_TO_PROCEDURE_CALL,
+                errorContext,
+                internal = false)
+        }
+        val bindingName = BindingName(arg.stringValue(), BindingCase.INSENSITIVE)
+        return when(val value = env.current[bindingName]) {
+            null -> valueFactory.missingValue
+            else -> value
+        }
+    }
+}
```

Comment:
This sproc is super interesting, however I don't think that we should support accessing the local variables of the current context this in this fashion because `Environment` is full of implementation details of the interpreter and that might be hard to support if we later want to change how the interpreter works in some substantive way, or say, replace the interpreter entirely (likely).

In looking at how the `env` parameter is actually used in all the built-in `ExprFunction` implementations, it's only ever used by `utcnow` to obtain the current `EvaluationSession`.  So I think we should change the `env: Environment` argument of `ExprFunction.call` to `session: EvaluationSession`, but that is out of scope for now.

Anyway, this makes me think we're better off creating a new interface so we can avoid spreading that problem around for now.  There will be a little duplication here but I think that's better than 

Possibly, something like this:

```
data class StoredProcedureSignature(val name: String, val arity: IntRange)
interface StoredProcedure {
    val signature: StoredProcedureSignature
    fun call(args: List<ExprValue>, session: EvaluationSession, valueFactory: ExprValueFactory) 
}
```

We throw `valueFactory: ExprValueFactory` in there so `StoredProcedure` implementations can avoid having to inject their own `ExprValueFactory` instances.  Let me know if you see any downsides to this.

Reply:
Added the `StoredProcedure` interface that takes in an `EvaluationSession` rather than an `Environment`. 

In the latest revision, I decided to leave out `ExprValueFactory` in the `call()`. I felt this is more in line with what `ExprFunction` currently does, where function implementations are passed an `ExprValueFactory` when created. Let me know if we shouldn't follow `ExprFunction`'s way of passing an `ExprValueFactory`.